Expell net::packet from output_stream API stack

xemul opened the pull request:
The existing data_sink_impl API has three put() virtual overloads

* put(net::packet) -- pure virtual one
* put(temporary_buffer) -- implementation creates net::packet out of the buffer
* put(vector<temporary_buffer>) -- similarly, the implementation converts the vector to packet

Also there's fallback_put(net::packet) method for those implementations that don't want to mess with packet and prefer to convert the packet back to buffers.

This API presumably was driven by the output_stream zero-copy buffers extension, that maintains net::packet on the output stream itself that accumulates written buffers and then put()-s them into sink. And the fallback_put() appeared later to facilitate sink implementations.

Maintaining net::packet as zero-copy buffers on output_stream is, in turn, makes the stream work in two (and a half) modes -- the users can either buffers data, then flush, or append zero-copy buffers, then, again, flush. There's also a semi-mixed mode, where zero-copy buffers may come after a bunch of buffered writes. Using mixed mode should still happen with care -- after zero-copy writes and flush(), starting buffered writes can step on assertion if the stream is batch-flushed.

Also the need to implement put(net::packet) overload is pretty harsh requirement, sinks that are not network sockets plug this implementation with abort() and require callers not to perform zero-copy writes into such streams.


This PR eliminates the net::packet from the output_stream+data_sink layer and leaves it on socket sink implementations only. For that both, output_stream and data_sink are changed.

First, the data_sink_impl. The new API (backward incompatible and thus under new API level) has just one put() that accepts std::span<temporary_buffer>. It's pure virtual method, implementations must grab the buffers before returning (even if the returned future is unresolved).

The data_sink() has the put(span) overload as well as put(temporary_buffer) and put(vector<temporary_buffer>) ones, for convenience.

File sink benefit from that change by just dropping the plugged put(packet) overload and that's mostly it. To submit buffers from span it picks them one-by-one, but later it should be tuned to submit iovec request.

Network sinks immediatly convert span of buffers into net::packet, no other changes are made.

HTTP chunked encoding sink converts each buffer from span into a "chunk", no other changes from current implementation is made. When we have mixed-mode (not this PR) it will be able to relax it to zero-copy buffers. The content-length sink zero-copy forwards the span to lower output_stream with the help of ... (see next paragraph)

The output_stream change is -- the _zc_bufs member is changed from net::packet to be std::vector<temporary_buffer> and zero-copy write()-s accepting packet and scattered_message are removed in the new API level (could be deprecated, but it would delay full expelling of packet from streams API). The write of single temporary buffer is preserved. A new write(span<temporary_buffer>) overload is added.

After the PR the existing _buf for buffered writes and _zc_bufs with zero-copied buffers still co-exist on output_stream and the old non-mixed behavior is preserved. However, this opens a way to implement the fully-mixed mode itself  :crossed_fingers:  eventually.

xemul commented:
@avikivity , please, share your thoughts about the general idea

avikivity commented:
> @avikivity , please, share your thoughts about the general idea

From reading the cover letter I'm in favor. For @nyh (also should be in the cover letter), net::packet is a low-level API representing packets for device drivers, having things like headroom and offload information, that are irrelevant for the higher levels of the stack. A net::packet sent from a high level will need to be split and merged in a tcp stream to suit the hardware and window limits.

avikivity commented on include/seastar/websocket/common.hh:
```diff
@@ -92,10 +92,16 @@ protected:
+#else
+        virtual future<> put(temporary_buffer<char> buf) override {
+            return data->push_eventually(std::move(buf));
+        }
```
Something like push_eventually will need to stabilize a span if it gets one.

avikivity replied:
Perhaps a better approach is to push immediately, but resolve the future only when control flow allows. So at least we know a fast implementation is possible.

avikivity commented on src/net/native-stack-impl.hh:
```diff
@@ -206,11 +206,28 @@ public:
+
+    virtual future<> put(std::vector<temporary_buffer<char>> bufs) override {
+        return put(net::packet::make(std::move(bufs)));
+    }
```
(the queue now is a vector of vectors because packet is fragmented)

avikivity replied:
Let's say the workload pushes 300-byte messages (and flushes the output stream after every message). We'll have a series of 300 byte packets (each with its own metadata header + headroom). Then tcp is awakened by a poller, gathers 1500 bytes - tcp overhead into a new packet, throws away the original packet headers. Wasteful.

With a queue of temporary_buffer:s, the tcp code can do

```c++
    if (!q.empty() && adjacent(q.back(), incoming_buffer)) {
         // merge incoming_buffer into q.back()
    } else {
         q.push_back(incoming_buffer);
    }
```

This reduces fragmentation (we observed that drivers have a hard time sending packets with small fragments and @gleb-cloudius even wrote code to linearize them). Even without  the merging, there's just less overhead.

avikivity commented on src/net/packet.cc:
```diff
@@ -137,6 +137,14 @@ std::ostream& operator<<(std::ostream& os, const packet& p) {
+    packet p(bufs.size());
+    for (auto&& buf : bufs) {
+        p = packet(std::move(p), std::move(buf));
+    }
```
Is there no nicer way?

xemul replied:
Not today :D There's a bunch of "appending" constructors for packet:

```c++
    // append fragment (copying new fragment)
    packet(packet&& x, fragment frag);
    // append fragment (zero-copy)
    packet(packet&& x, fragment frag, deleter d);
    // append temporary_buffer (zero-copy)
    packet(packet&& x, temporary_buffer<char> buf);
```

The `packet(std::vector<temporary_buffer>)` one (or better `packet(std::span<temporary_buffer>)`) in the next version TODO list

avikivity replied:
ok

xemul replied:
`net::packet(std::span<temporary_buffer>)` constructor in new version

xemul commented:
> > First, the data_sink. The new API (backward incompatible and thus under new API level) is two put()-s
> > 
> > * put(temporary_buffer) -- pure virtual one
> > * put(vector<temporary_buffer>) -- the implementation falls-back to putting vector elements with the former overload one-by-one
> 
> Should we accept a span here?

What do you call "a span"? Presumably the std::span<>. do you?

> I think the implementation will not be able to reuse the vector. If it has a queue of fragments, it won't be a queue of vectors of fragments, but a direct queue of fragments.
> 
> An advantages is that put(temporary_buffer) can call into put(span) and we only have one virtual entry point.

But how? Right now calling put(temporary_buffer) carries the buffer from output_stream into the data_sink. Calling put(span) requires that the buffer itself is stored somewhere, and this "somewhere" shouldn't be the output_stream any longer, it should be std::move()-d from there.

> > If the sink implementation thinks it can do vector put better, than sequentially it can override one and this is what socket sinks do -- they convert the passed vector of buffers into net::packet and proceed. In fact, this is what already happens today, partially at the output_stream level itself, so this change just moves the net::packet instantiation to src/net/ code.
> > File sink benefit from that change by just dropping the plugged put(packet) overload and that's it.
> > The loser here is http content length body sink that stops being zero-copy, but that'll be changed in v2 (if it follows).
> 
> Why does it stop being zero-copy?

It's mostly a note for myself not to forget and fix that in vN+1

Before the patch it had:

* put(packet)  -> zero-copy move the packet into underlying output_stream::write(packet)
* put(temporary_buffer) -> buffered copy the buffer into underlying output_stream::write(buffer.get(), buffer.size())

And this PR just drops the former put() and leaves the latter. One more patch and it's zero-copy again.

> > The output_stream change is -- the _zc_bufs member is changed from net::packet to be std::vector<temporary_buffer>
> 
> Excellent.
> 
> > and zero-copy write()-s accepting packet and scattered_message are deprecated in favor of (exiting) write of temporary_buffer itself.
> > The deprecated write()-s are implemented to convert packet/scattered_message into a vector of buffers and write the buffers themselves. That's going to generate some CPU overhead converting things back-and-forth plus memory allocation, but these methods are deprecated and callers are support to switch to write(temporary_buffer) which doesn't do it.
> 
> span will eliminate the memory allocation for the small cases.

Smallest case (vector with one buffer) doesn't allocate more than the vector storage itself in this PR.
But again, spam is not a container, so whether or not it will eliminate allocations depends on the underlying container. What am I missing?

> > Right now the existing _buf for buffered writes and _zc_bufs with zero-copied buffers co-exist on output_stream and the old non-mixed behavior is preserved. However, next step is the fully-mixed mode itself 🤞

xemul commented on src/net/posix-stack.cc:
```diff
@@ -681,13 +681,23 @@ posix_data_sink_impl::put(temporary_buffer<char> buf) {
+    _p = net::packet::make(std::move(bufs));
+    auto sg_id = internal::scheduling_group_index(current_scheduling_group());
+    bytes_sent[sg_id] += _p.len();
+    return _fd.write_all(_p).then([this] { _p.reset(); });
```
It's pollable_fd::write_all(packet) that instantly maps into sendmsg, what's wrong with using packet at that level?

xemul commented on include/seastar/core/iostream.hh:
```diff
@@ -447,9 +447,10 @@ class output_stream final {
     data_sink _fd;
     temporary_buffer<CharType> _buf;
-    net::packet _zc_bufs = net::packet::make_null_packet(); //zero copy buffers
+    std::vector<temporary_buffer<CharType>> _zc_bufs; // zero copy buffers
```
> Maybe circular_buffer?

I thought of it. It only makes life better in zero_copy_split_and_put(). But [there are](https://github.com/scylladb/seastar/pull/2937#discussion_r2291138302) other ways to handle it.

> btw, do we have flow control over _zc_bufs, or can it grow indefinitely?

It can grow infinitely, but once its total length hits the stream buffer size it's being put() into sink.

It's the same as constructing the net::packet outside of output_stream before write()-ing it today.
E.g.  [`cql_server::response::make_message`](https://github.com/scylladb/scylladb/blob/47750496d21a8260f896f3d203bd50f4ce4135b7/transport/server.cc#L1705-L1716). If it's patched to use this PR's recommended output_stream API, it would grow _zc_bufs as large as it nowadays grows the net::packet.

avikivity commented:
> > Yes, data_sink_derived_class::put(std::span<temporary_buffer>) should move all its input buffers.
> 
> Well, I had a version that only had data_sink::put(small_vector<temporary_buffer, 1>), so single buffer vs multiple buffers is all the same, and allocation optimization for single-buffer case. It turned out that all non-network sink implementations are duplicating their code like this
> 
> ```
>     private future<> put(temporary_buffer) {
>         // single buffer processing
>     }
>     public future<> put(small_vector<temporary_buffer> buffers) override {
>         for (auto& buf : buffers) {
>             co_await put(std::move(buf));
>         }
>     }
> ```
> 
> Adding virtual single-buffer put() is surprisingly less code. I think it will be the same for spans.
> 

Ok. But the naive loop is wrong, it's not letting the lower layer coalesce. For example, posix tcp can use a single sendmsg instead of multiple ones.

I think that's some motivation for removing the non-span version - to discourage lazy+bad implementations.


> > > > span will eliminate the memory allocation for the small cases.
> > > 
> > > 
> > > Smallest case (vector with one buffer) doesn't allocate more than the vector storage itself in this PR. But again, spam is not a container, so whether or not it will eliminate allocations depends on the underlying container. What am I missing?
> > 
> > 
> > The allocations of the vector itself? Not comparing to the patches, but to existing state.
> 
> Yes. The deprecated write(net::packet) doesn't allocate a vector of temporary_buffers is the packet consists of a single fragment. If comparing to existing state it becomes worse -- before the patches, a write(net::packet) into an empty stream just moves the packet onboard, after the patches a write(net::packet) into an empty stream allocates an inner vector storage to hold temporary_vector constructed out of the packet's buffers

We shouldn't compare to write(net::packet) since we're deprecating it.

Perhaps we should add write(span\<temporary_buffer>).

avikivity commented on include/seastar/core/temporary_buffer.hh:
```diff
@@ -32,6 +32,7 @@
 #include <seastar/util/std-compat.hh>
 #include <seastar/util/modules.hh>
+#include <seastar/util/assert.hh>
 
```
```c++
        future<> respond(udp_channel& chan) {
            int i = 0;
            return do_for_each(_out_bufs.begin(), _out_bufs.end(), [this, i, &chan] (packet& p) mutable {
                header* out_hdr = p.prepend_header<header>(0);
                out_hdr->_request_id = _request_id;
                out_hdr->_sequence_number = i++;
                out_hdr->_n = _out_bufs.size();
                *out_hdr = hton(*out_hdr);
                return chan.send(_src, std::move(p));
            });
        }
```

We can keep trim_to_size, I see it helps http.

xemul commented on include/seastar/core/iostream.hh:
```diff
@@ -495,8 +495,10 @@ public:
     future<> write(net::packet p) noexcept;
     /// Appends the scattered message as zero-copy buffer
+    [[deprecated("Use write(temporary_buffer) overload")]]
     future<> write(scattered_message<char_type> msg) noexcept;
```
Removed this overload for API_LEVEL >= 9 in new version

xemul commented:
upd:
Keep chunked-encoded http sink use buffered writes. Filly-mixed mode for output_stream is required

xemul commented:
upd:
- do not append empty buffers to zc-bufs vector

avikivity commented on include/seastar/net/packet.hh:
```diff
@@ -228,6 +229,8 @@ public:
     // append deleter
     packet(packet&& x, deleter d);
+    // create from span of buffers (zero-copy, ownership is transfered)
+    packet(std::span<temporary_buffer<char>>);
```
explicit

avikivity commented on include/seastar/core/iostream.hh:
```diff
@@ -151,6 +159,23 @@ public:
+
+        auto f = fn(std::move(bufs.front()));
+        for (auto&& buf : bufs.subspan(1)) {
+            f = std::move(f).then([fn, buf = std::move(buf)] () mutable {
```
Doesn't this violate the contract? Nothing holds bufs safe.

I'd like the name of the function to better help the caller understand when to call it.

avikivity replied:
Ah, the name already exists.

avikivity replied:
Well, it should inherit the comment.

avikivity replied:
> Doesn't this violate the contract? Nothing holds bufs safe.

It does not violate the contract, it creates a large future chain.

avikivity replied:
Can be written as: std::ranges::fold_left(bufs | std::views::as_rvalue, make_ready_future<>(), []  (future<>, temporary_buffer) { ... });

xemul replied:
Yes, but only after we stop supporting C++20 :man_shrugging:

avikivity commented on include/seastar/websocket/common.hh:
```diff
@@ -92,10 +92,18 @@ protected:
+#if SEASTAR_API_LEVEL >= 9
+        future<> put(std::span<temporary_buffer<char>> d) override {
+            return data_sink_impl::fallback_put(d, [this] (temporary_buffer<char>&& buf) {
+                return data->push_eventually(std::move(buf));
```
Note: now I think it's bad design to use an extra queue here. There should be one queue, tcp. But not the fault of this series.

avikivity commented:
Looks good. But title needs to indicate the transition from net::packet to span\<t_b>, and cover letter needs updating.

xemul commented on include/seastar/core/iostream.hh:
```diff
@@ -472,6 +518,8 @@ public:
     /// Appends the temporary buffer as zero-copy buffer
     future<> write(temporary_buffer<char_type>) noexcept;
+    /// Appends a bunch of buffers as zero-copy
+    future<> write(std::span<temporary_buffer<char_type>>) noexcept;
```
This write overload is needed for http content-length data sink which in turn needs to forward span of buffers from its put() method down to another _output_stream_, not to other sink. I'll probably re-split this PR a bit differently to address that

xemul commented on include/seastar/core/temporary_buffer.hh:
```diff
@@ -249,4 +250,34 @@ public:
+namespace internal {
+
+template <typename CharType>
+inline std::vector<temporary_buffer<CharType>> split_buffers(std::vector<temporary_buffer<CharType>>& bufs, size_t offset) {
```
Renamed it to `detach_front()` but it still updates the input vector "in place". The something-returning option doesn't look better, but we can change it later.

And added documentation

xemul commented:
upd:
- renamed split_buffers -> detach_front and documented it
- introduced output_stream::write(span<temporary_buffer>) with separate patch
- changed the way write(packet) converts packet to span of buffers
